cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java

```diff
@@ -71,6 +77,103 @@ public Schema getSchema() {
+   * logical type date
+   */
+  public LocalDate getDate(String fieldName) {
+    if (!isValidLogicalType(fieldName, Schema.LogicalType.DATE)) {
+      return null;
+    }
+
+    return LocalDate.ofEpochDay(((Integer) fields.get(fieldName)).longValue());
+  }
+
+  /**
+   * Get {@link LocalTime} from field of logical type time.
+   * @param fieldName time field to get.
+   * @return value of the field as an object of {@link LocalTime}, returns null if given field is not of
+   * logical type time
+   */
+  public LocalTime getTime(String fieldName) {
+    if (!isValidLogicalType(fieldName, Schema.LogicalType.TIME_MICROS)
+      && !isValidLogicalType(fieldName, Schema.LogicalType.TIME_MILLIS)) {
+      return null;
+    }
+
+    Schema fieldSchema = this.schema.getField(fieldName).getSchema();
+    if (fieldSchema.getLogicalType().equals(Schema.LogicalType.TIME_MILLIS)) {
+      return LocalTime.ofNanoOfDay(TimeUnit.MILLISECONDS.toNanos(((Integer) fields.get(fieldName))));
+    }
+
+    return LocalTime.ofNanoOfDay(TimeUnit.MICROSECONDS.toNanos((Long) fields.get(fieldName)));
+  }
+
+  /**
+   * Get {@link ZonedDateTime} from field of logical type timestamp. Default {@link ZoneId} is system default.
+   * @param fieldName zoned date time field to get.
+   * @return value of the field as an object of {@link ZonedDateTime}, returns null if given field is not of
+   * logical type timestamp
+   */
+  public ZonedDateTime getTimestamp(String fieldName) {
+    if (!isValidLogicalType(fieldName, Schema.LogicalType.TIMESTAMP_MILLIS)
+      && !isValidLogicalType(fieldName, Schema.LogicalType.TIMESTAMP_MICROS)) {
+      return null;
+    }
+    return getZonedDateTime(fieldName, ZoneId.systemDefault());
+  }
+
+  /**
+   * Get {@link ZonedDateTime} from field of logical type timestamp.
+   * @param fieldName zoned date time field to get.
+   * @param zoneId zone id for the field
+   * @return value of the field as an object of {@link ZonedDateTime}
+   */
+  public ZonedDateTime getTimestamp(String fieldName, ZoneId zoneId) {
+    if (!isValidLogicalType(fieldName, Schema.LogicalType.TIMESTAMP_MILLIS)
+      && !isValidLogicalType(fieldName, Schema.LogicalType.TIMESTAMP_MICROS)) {
+      return null;
+    }
+
+    return getZonedDateTime(fieldName, zoneId);
+  }
+
+  /**
+   * Get zoned date and time represented by the field.
+   * @param fieldName Name of the field
+   * @param zoneId zone id for the field
+   * @return {@link ZonedDateTime} represented by field.
+   */
+  private ZonedDateTime getZonedDateTime(String fieldName, ZoneId zoneId) {
+    Long ts = (Long) fields.get(fieldName);
+    // get fractional portion of timestamp, we always store timestamp in micro seconds, so first get that portion.
+    // for example: timestamp is 1970-01-01T00:00:01.112456Z, here fraction = 112456, tsInSeconds = 1
+    int fraction = (int) (ts % (1000000L));
+    long tsInSeconds = (ts - fraction) / (1000000L);
```

Comment:
Why need the subtraction? I think just the simple integral division would do it.